File: PoketDex/app/src/main/java/pt/rfsfernandes/ui/fragments/details_pager_fragments/GeneralInfoFragment.java
package pt.rfsfernandes.ui.fragments.details_pager_fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import pt.rfsfernandes.MyApplication;
import pt.rfsfernandes.custom.Constants;
import pt.rfsfernandes.custom.adapters.PokemonTypesAdapter;
import pt.rfsfernandes.databinding.FragmentGeneralInfoBinding;
import pt.rfsfernandes.model.pokemon.Pokemon;
import pt.rfsfernandes.ui.activities.MainActivity;
import pt.rfsfernandes.viewmodels.MainViewModel;


public class GeneralInfoFragment extends Fragment {
  private FragmentGeneralInfoBinding binding;
  private MainViewModel mMainViewModel;
  private MyApplication mMyApplication;
  private PokemonTypesAdapter mPokemonTypesAdapter;
  private boolean isGeneralInfoLoaded = false;
  private boolean isDescriptionLoaded = false;

  public GeneralInfoFragment() {
    // Required empty public constructor
  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
                           Bundle savedInstanceState) {

    binding = FragmentGeneralInfoBinding.inflate(inflater, container, false);
    View view = binding.getRoot();
    mMainViewModel = new ViewModelProvider(requireActivity()).get(MainViewModel.class);
    mMyApplication = (MyApplication) requireActivity().getApplication();
    mPokemonTypesAdapter = new PokemonTypesAdapter(requireContext());
    binding.pokemonTypeGridView.setAdapter(mPokemonTypesAdapter);
    return view;
  }

  @Override
  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);
    initViewModel();

    binding.pokemonTypeGridView.setOnItemClickListener((parent, view1, position, id) -> {
      mMainViewModel.isLoading(true);
      if (getActivity() != null) {
        ((MainActivity) getActivity()).setShowTypeInfo(true);
        mMainViewModel.getTypeAndCounters((int) id, Constants.SHOW_TYPE.POKEMON);
      }
    });

  }

  /**
   * Initiates viewModel observers
   */
  private void initViewModel() {

    mMainViewModel.getPokemonMutableLiveData().observe(getViewLifecycleOwner(), pokemon -> {
      if (pokemon != null) {

        populateGeneralViews(pokemon);
        inflatePokemonTypes(pokemon);

        isGeneralInfoLoaded = true;
        stopLoading();

      }
    });

    mMainViewModel.getPokemonDescriptionLiveData().observe(getViewLifecycleOwner(),
        description -> {
          binding.textViewDescription.setText(description);
          isDescriptionLoaded = true;
          stopLoading();
        });

  }

  /**
   * Stops the progressBar loading
   */
  private void stopLoading() {

    if (isDescriptionLoaded && isGeneralInfoLoaded) {
      mMainViewModel.isLoading(false);
      isDescriptionLoaded = false;
      isGeneralInfoLoaded = false;
    }

  }

  /**
   * Populates general views such as images and pokemon name
   *
   * @param pokemon selected pokemon
   */
  private void populateGeneralViews(Pokemon pokemon) {

    binding.imageViewPokemonDefault.setImageURI(pokemon.getSprites().getFrontDefault());
    binding.imageViewPokemonShiny.setImageURI(pokemon.getSprites().getFrontShiny());

  }

  /**
   * Inflates linearLayoutTypeContainer dynamically with pokemon types
   *
   * @param pokemon selected pokemon
   */
  private void inflatePokemonTypes(Pokemon pokemon) {
    mPokemonTypesAdapter.refreshList(pokemon.getTypeList());
  }
}

File: PoketDex/app/src/main/java/pt/rfsfernandes/pocketdex/ui/fragments/details_pager_fragments/GeneralInfoFragment.kt
package pt.rfsfernandes.pocketdex.ui.fragments.details_pager_fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import androidx.fragment.app.Fragment
import org.koin.androidx.viewmodel.ext.android.sharedViewModel
import pt.rfsfernandes.pocketdex.custom.Constants
import pt.rfsfernandes.pocketdex.custom.adapters.PokemonTypesAdapter
import pt.rfsfernandes.pocketdex.databinding.FragmentGeneralInfoBinding
import pt.rfsfernandes.pocketdex.MyCustomApplication
import pt.rfsfernandes.pocketdex.model.pokemon.Pokemon
import pt.rfsfernandes.pocketdex.ui.activities.MainActivity
import pt.rfsfernandes.pocketdex.viewmodels.MainViewModel

class GeneralInfoFragment : Fragment() {
    private lateinit var binding: FragmentGeneralInfoBinding
    private val mMainViewModel: MainViewModel by sharedViewModel()
    private lateinit var mMyCustomApplication: MyCustomApplication
    private lateinit var mPokemonTypesAdapter: PokemonTypesAdapter
    private var isGeneralInfoLoaded = false
    private var isDescriptionLoaded = false
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        binding = FragmentGeneralInfoBinding.inflate(inflater, container, false)
        val view: View = binding.root
        mMyCustomApplication = requireActivity().application as MyCustomApplication
        mPokemonTypesAdapter = PokemonTypesAdapter(requireContext())
        binding.pokemonTypeGridView.adapter = mPokemonTypesAdapter
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViewModel()
        binding.pokemonTypeGridView.onItemClickListener = OnItemClickListener { parent: AdapterView<*>?, view1: View?, position: Int, id: Long ->
            mMainViewModel.isLoading(true)
            if (activity != null) {
                (activity as MainActivity?)!!.setShowTypeInfo(true)
                mMainViewModel.getTypeAndCounters(id.toInt(), Constants.SHOW_TYPE.POKEMON)
            }
        }
    }

    /**
     * Initiates viewModel observers
     */
    private fun initViewModel() {
        mMainViewModel.pokemonMutableLiveData.observe(viewLifecycleOwner) { pokemon: Pokemon? ->
            if (pokemon != null) {
                populateGeneralViews(pokemon)
                inflatePokemonTypes(pokemon)
                isGeneralInfoLoaded = true
                stopLoading()
            }
        }
        mMainViewModel.pokemonDescriptionLiveData.observe(viewLifecycleOwner
        ) { description: String? ->
            binding.textViewDescription.text = description
            isDescriptionLoaded = true
            stopLoading()
        }
    }

    /**
     * Stops the progressBar loading
     */
    private fun stopLoading() {
        if (isDescriptionLoaded && isGeneralInfoLoaded) {
            mMainViewModel.isLoading(false)
            isDescriptionLoaded = false
            isGeneralInfoLoaded = false
        }
    }

    /**
     * Populates general views such as images and pokemon name
     *
     * @param pokemon selected pokemon
     */
    private fun populateGeneralViews(pokemon: Pokemon) {
        binding.imageViewPokemonDefault.setImageURI(pokemon.sprites.frontDefault)
        binding.imageViewPokemonShiny.setImageURI(pokemon.sprites.frontShiny)
    }

    /**
     * Inflates linearLayoutTypeContainer dynamically with pokemon types
     *
     * @param pokemon selected pokemon
     */
    private fun inflatePokemonTypes(pokemon: Pokemon) {
        mPokemonTypesAdapter.refreshList(pokemon.typeList)
    }
}